dot/parachain/candidate-validation/worker_pool.go

```diff
@@ -0,0 +1,165 @@
+	CandidateCommitments    parachaintypes.CandidateCommitments
+	PersistedValidationData parachaintypes.PersistedValidationData
+}
+
+type ReasonForInvalidity byte
+
+const (
+	// ExecutionError Failed to execute `validate_block`. This includes function panicking.
+	ExecutionError ReasonForInvalidity = iota
+	// InvalidOutputs Validation outputs check doesn't pass.
+	InvalidOutputs
+	// Timeout Execution timeout.
+	Timeout
+	// ParamsTooLarge Validation input is over the limit.
+	ParamsTooLarge
+	// CodeTooLarge Code size is over the limit.
+	CodeTooLarge
+	// PoVDecompressionFailure PoV does not decompress correctly.
+	PoVDecompressionFailure
+	// BadReturn Validation function returned invalid data.
+	BadReturn
+	// BadParent Invalid relay chain parent.
+	BadParent
+	// PoVHashMismatch POV hash does not match.
+	PoVHashMismatch
+	// BadSignature Bad collator signature.
+	BadSignature
+	// ParaHeadHashMismatch Para head hash does not match.
+	ParaHeadHashMismatch
+	// CodeHashMismatch Validation code hash does not match.
+	CodeHashMismatch
+	// CommitmentsHashMismatch Validation has generated different candidate commitments.
+	CommitmentsHashMismatch
+)
+
+func (ci ReasonForInvalidity) Error() string {
+	switch ci {
+	case ExecutionError:
+		return "failed to execute `validate_block`"
+	case InvalidOutputs:
+		return "validation outputs check doesn't pass"
+	case Timeout:
+		return "execution timeout"
+	case ParamsTooLarge:
+		return "validation input is over the limit"
+	case CodeTooLarge:
+		return "code size is over the limit"
+	case PoVDecompressionFailure:
+		return "PoV does not decompress correctly"
+	case BadReturn:
+		return "validation function returned invalid data"
+	case BadParent:
+		return "invalid relay chain parent"
+	case PoVHashMismatch:
+		return "PoV hash does not match"
+	case BadSignature:
+		return "bad collator signature"
+	case ParaHeadHashMismatch:
+		return "para head hash does not match"
+	case CodeHashMismatch:
+		return "validation code hash does not match"
+	case CommitmentsHashMismatch:
+		return "validation has generated different candidate commitments"
+	default:
+		return "unknown invalidity reason"
+	}
+}
+
+func newValidationWorkerPool() *workerPool {
+	return &workerPool{
+		workers: make(map[parachaintypes.ValidationCodeHash]*worker),
+	}
+}
+
+func (v *workerPool) newValidationWorker(validationCode parachaintypes.ValidationCode) error {
+
+	worker, err := newWorker(validationCode)
+	if err != nil {
+		return fmt.Errorf("failed to create a new worker: %w", err)
+	}
+
+	v.workers[worker.workerID] = worker
+
+	return nil
+}
+
+// submitRequest given a request, the worker pool will get the worker for a given task and submit the request
+// to the worker. The worker will execute the request and return the result. If the worker does not exist, a new worker
+// will be created and the request will be submitted to the worker.
+func (v *workerPool) submitRequest(msg *ValidationTask) (*ValidationResult, error) {
```

Comment:
```suggestion
func (v *workerPool) executeRequest(msg *ValidationTask) (*ValidationResult, error) {
```
Maybe mirror the name since it's not submitting to a pool anymore, but rather will be executed synchronously.

Reply:
Good point, updated.